### DISCUSSION TITLE
support for “no-responders” in multitenancy

### DISCUSSION THREAD

**Discussion by piotrkwarcinski** [nats-io/nats-server#4761] [Q&A] [answered]
Hi, I've got a question regarding multi tenancy and [no-responders](https://docs.nats.io/nats-concepts/core-nats/queue#no-responder). I defined the following simplified configuration:
```
accounts: {
  main: {
    users: [{user: main, password: a}]
    imports: [{service: {subject: agents.>, account: client}, to: client.agents.>}]
  },
  client: {
    users: [{user: client, password: b}]
    exports: [{service: agents.>, accounts: [main]}]
  },
}
```
And the following client:
```
package main

import (
	"fmt"
	"time"

	"github.com/nats-io/nats.go"
)

func main() {
	mainURL := "nats://main:a@localhost:4222"

	mn, _ := nats.Connect(mainURL)
	defer mn.Drain()

	resp, err := mn.Request("client.agents.Sub", nil, time.Second)
	if err != nil {
		panic(err)
	} else {
		fmt.Printf("subject: %s, data: %s\n", resp.Subject, resp.Data)
	}
}
```
The problem is that I receive a timeout ` nats: timeout` instead of expected `nats: no responders available for request`. I think the problem is with the exported subject because when I request a message to an unexported subject `not_agents`:
```
package main  
  
import (  
    "fmt"  
    "time"  
    "github.com/nats-io/nats.go")  
  
func main() {  
    mainURL := "nats://main:a@localhost:4222"  
  
    mn, _ := nats.Connect(mainURL)  
    defer mn.Drain()  
  
    resp, err := mn.Request("not_agents", nil, time.Second)  
    if err != nil {  
       panic(err)  
    } else {  
       fmt.Printf("subject: %s, data: %s\n", resp.Subject, resp.Data)  
    }  
}
```
I get ` nats: no responders available for request` immediately which is the expected behavior.

I tried to debug this a little and it seems that when a subject is imported an internal (inside nats-server) nats subscription is created on this subject  so technically because of this internal subscription there is an available responder and that's why I get a timeout instead of `no responders available`.

Could you confirm that  “no-responders”  is not supported when communicating across accounts using multi tenancy? Do you know about any workaround which can cause to return `no responders available for request` immediately (without timeout) using a similar configuration (with exported/imported subjects)?

**Comment by jnmoyne** [comment, marked as answer]
As you figured out technically there is a listener on the subject that is created under the covers due to the import of the service. This means that indeed when it comes to cross-accounts services import/export you can not rely on "no responders" and have to resort to wait for a timeout. 

Note that "no responders" is a kind of "shortcut" to not receiving an error but even when it is available the invocation of the service can still time out. So in your client application code you should expect that both no responders and timeouts can happen and should be handled the same way.
